compiler/noirc_evaluator/src/ssa/ssa_gen/mod.rs

```diff
@@ -752,6 +754,155 @@ impl<'a> FunctionContext<'a> {
         })
     }
 
+    fn codegen_match(&mut self, match_expr: &ast::Match) -> Result<Values, RuntimeError> {
+        let variable = self.lookup(match_expr.variable_to_match);
+
+        // Any matches with only a single case we don't need to check the tag at all.
+        // Note that this includes all matches on struct / tuple values.
+        if match_expr.cases.len() == 1 && match_expr.default_case.is_none() {
+            return self.no_match(variable, &match_expr.cases[0]);
+        }
+
+        // From here on we can assume `variable` is an enum, int, or bool value (not a struct/tuple)
+        let tag = self.enum_tag(&variable);
+        let tag_type = self.builder.type_of_value(tag).unwrap_numeric();
+
+        let end_block = self.builder.insert_block();
+
+        // Optimization: if there is no default case we can jump directly to the last case
+        // when finished with the previous case instead of using a jmpif with an unreachable
+        // else block.
+        let last_case = if match_expr.default_case.is_some() {
+            match_expr.cases.len()
+        } else {
+            match_expr.cases.len() - 1
+        };
+
+        for i in 0..last_case {
+            let case = &match_expr.cases[i];
+            let variant_tag = self.variant_index_value(&case.constructor, tag_type)?;
+            let eq = self.builder.insert_binary(tag, BinaryOp::Eq, variant_tag);
+
+            let case_block = self.builder.insert_block();
+            let else_block = self.builder.insert_block();
+            self.builder.terminate_with_jmpif(eq, case_block, else_block);
+
+            self.builder.switch_to_block(case_block);
+            self.bind_case_arguments(variable.clone(), case);
+            let results = self.codegen_expression(&case.branch)?.into_value_list(self);
+            self.builder.terminate_with_jmp(end_block, results);
+
+            self.builder.switch_to_block(else_block);
+        }
+
+        if let Some(branch) = &match_expr.default_case {
+            let results = self.codegen_expression(branch)?.into_value_list(self);
+            self.builder.terminate_with_jmp(end_block, results);
+        } else {
+            // If there is no default case, assume we saved the last case from the
+            // last_case optimization above
+            let case = match_expr.cases.last().unwrap();
```

Comment:
We allow empty types right? Can this potentially panic if we had an empty match statement?

Reply:
That code works as expected - you can even illegally create such a type via `std::mem::zeroed::<Empty>()` and call `my_match`. When this happens the match contains one branch which is just `assert(false);`

Reply:
I think it would be a good test case still, but can be done in follow-ups as this is experimental